File: linkgame-server/src/main/java/io/github/octcarp/linkgame/server/net/LobbyManager.java
package io.github.octcarp.linkgame.server.net;

import io.github.octcarp.linkgame.common.packet.Response;
import io.github.octcarp.linkgame.common.packet.ResponseType;

import java.util.HashMap;
import java.util.Map;

public class LobbyManager {
    private static LobbyManager instance;

    private final Map<String, ClientHandlerThread> lobbyPlayerThreads;
    private final Map<String, String> waitingPlayers;

    private LobbyManager() {
        lobbyPlayerThreads = new HashMap<>();
        waitingPlayers = new HashMap<>();
    }

    public static LobbyManager getInstance() {
        if (instance == null) {
            instance = new LobbyManager();
        }
        return instance;
    }

    public ClientHandlerThread getClientThreadByPlayerId(String playerId) {
        return lobbyPlayerThreads.get(playerId);
    }

    public void enterLobby(String playerId, ClientHandlerThread clientHandlerThread) {
        lobbyPlayerThreads.put(playerId, clientHandlerThread);
        notifyAllLobbyPlayers();
    }

    public void exitLobby(String playerId) {
        lobbyPlayerThreads.remove(playerId);
        notifyAllLobbyPlayers();
    }

    public void notifyAllLobbyPlayers() {
        Response response = new Response(ResponseType.ALL_MATCHING_PLAYER, lobbyPlayerThreads.keySet());
        for (ClientHandlerThread clientHandlerThread : lobbyPlayerThreads.values()) {
            clientHandlerThread.sendResponse(response);
        }
    }

}

File: game-server/src/main/kotlin/io/github/octcarp/sustech/cs209a/linkgame/server/net/LobbyManager.kt
package io.github.octcarp.sustech.cs209a.linkgame.server.net

import io.github.octcarp.sustech.cs209a.linkgame.common.packet.Response
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.ResponseType
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentSkipListSet

object LobbyManager {
    private val lobbyPlayerThreads = ConcurrentHashMap<String, ClientHandlerThread>()
    private val onlinePlayers = ConcurrentSkipListSet<String>()
    private val waitingPlayers = ConcurrentSkipListSet<String>()
    private val exceptPlayers = ConcurrentHashMap<String, String>()

    fun getClientThreadByPlayerId(playerId: String): ClientHandlerThread? {
        return lobbyPlayerThreads[playerId]
    }

    fun enterLobby(playerId: String, clientHandlerThread: ClientHandlerThread) {
        lobbyPlayerThreads[playerId] = clientHandlerThread

        exceptPlayers[playerId]?.let { oppId ->
            val response = Response(ResponseType.WAITING_OPPONENT, oppId)
            clientHandlerThread.sendResponse(response)
        }

        val clientResponse = Response(ResponseType.ALL_WAITING_PLAYERS, waitingPlayers.toList())
        clientHandlerThread.sendResponse(clientResponse)
    }

    fun startWaiting(playerId: String) {
        if (waitingPlayers.add(playerId)) {
            notifyAllLobbyPlayers()
        }
    }

    fun stopWaiting(playerId: String) {
        if (waitingPlayers.remove(playerId)) {
            notifyAllLobbyPlayers()
        }
    }

    fun exitLobby(playerId: String) {
        lobbyPlayerThreads.remove(playerId)
        if (waitingPlayers.remove(playerId)) {
            notifyAllLobbyPlayers()
        }
    }

    fun notifyAllLobbyPlayers() {
        val response = Response(ResponseType.ALL_WAITING_PLAYERS, waitingPlayers.toList())
        lobbyPlayerThreads.values.forEach {
            it.sendResponse(response)
        }
    }

    fun joinPlayer(playerId: String, oppId: String) {
        lobbyPlayerThreads[playerId]?.let { client ->
            lobbyPlayerThreads[oppId]?.let { oppClient ->
                stopWaiting(playerId)
                stopWaiting(oppId)
                MatchManager.createMatch(oppId, playerId, oppClient, client)
            }
        }
    }


    fun playerDisconnected(playerId: String) {
        lobbyPlayerThreads.remove(playerId)
        if (waitingPlayers.remove(playerId)) {
            notifyAllLobbyPlayers()
        }
    }
}
